streamingest: replicate span configs relevant to replicating tenant

msbutler opened the pull request:
This patch spawns a new goroutine on the stream ingestion job coordinator which
in turn creates a special span config client and subscription for replicating
span config updates. As updates come in, the coordinator will buffer updates
and deletes with the same source side commit timestamp and write them in a
transaction to the destination side system span configuration table when it
observes a new update with a newer timestamp.

The ingestion side assumes each replicated update is unique and in timestamp
order, which is enforced by the producer side logic built in https://github.com/cockroachdb/cockroach/pull/108356. This
assumption simplifies the destination side ingestion logic which must write
updates with the same source side transaction commit timestamp at the
same new timestamp on the destination side. This invariant ensures a span
configuration's target (i.e. the span that a configuration applies to) never
overlaps with any other span configuration target. Else, c2c would break the
span config reconciliation system.

Note that cutover will not revert any ingested span config updates, as we
wouldn't not want to issue revert range requests on an online system table in
the system tenant. That being said, a future PR will need to teach the
destination side application tenant to conduct a full span reconcilation job
immediately after cutover, which will safely revert any span config updates
that committed after the cutover timestamp.

A future PR will also improve the ingestion logic during an initial scan on
Resume. Because the job does not maintain a protected timestamp on the span
config table, the job may miss updates if it resumes after a long pause period.
So, the ingestor will need to buffer all initial scan updates and flush them in
one transaction along with a tenant span wide delete, in order to maintain the
span config table invariants.

Informs https://github.com/cockroachdb/cockroach/issues/106823

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107557)
<!-- Reviewable:end -->

msbutler commented:
@stevendanna this is ready for a first pass! You've already reviewed the first commit.

stevendanna commented on pkg/ccl/streamingccl/streamclient/client.go:
```diff
@@ -161,6 +161,9 @@ func NewStreamClient(
 	case "postgres", "postgresql":
 		// The canonical PostgreSQL URL scheme is "postgresql", however our
 		// own client commands also accept "postgres".
+		if processOptions(opts).forSpanConfigs {
+			return NewSpanConfigStreamClient(ctx, streamURL, opts...)
+		}
```
[nit] I would mildly prefer that the caller just call NewSpanConfigStreamClient directly. That sets us up in the right place for it being its own type in the future.

If we leave it as is, I think we need to at least update the `external` branch of this switch statement to make sure it it passing the opts through.

stevendanna commented on pkg/ccl/streamingccl/streamingest/ingest_span_configs.go:
```diff
@@ -0,0 +1,253 @@
+
+	g := ctxgroup.WithContext(ctx)
+	g.GoCtx(sub.Subscribe)
+	g.GoCtx(sc.checkForCutover)
```
Do we need this checkForCutover?  Could we instead have the goroutine running the distsql flow shut us down when it exits, in the same way it shuts down the replanner?

msbutler replied:
Good point. Refactored to remove this.

stevendanna commented on pkg/ccl/streamingccl/streamingest/ingest_span_configs.go:
Why do we only check the length of bufferedUpdates here and not also the length of bufferedDeletes?

stevendanna commented on pkg/ccl/streamingccl/streamingest/ingest_span_configs.go:
Given the TODO, perhaps let's log what we are skipping for the time being, just in case we end up having to debug something around this?

stevendanna commented on pkg/ccl/streamingccl/streamingest/ingest_span_configs.go:
Consider moving this to 1 or 2.  I think vlog > 2 is usually for the noisiest stuff.

stevendanna commented on pkg/ccl/streamingccl/streamingest/ingest_span_configs.go:
Consider adding a comment directing users to updateSpanConfigRecords for comparison.

stevendanna commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_dist.go:
```diff
@@ -185,12 +201,26 @@ func startDistIngestion(
+// TODO(msbutler): remove once spanConfig API is refactored. We need a way to
+// grab the sourceTenantID during distSQL planning. Perhaps we ought to store
+// the sourceTenantID id the job record.
+func (p *replicationFlowPlanner) getSrcTenantID() (roachpb.TenantID, error) {
+	if p.srcTenantID.InternalValue == 0 {
+		return p.srcTenantID, errors.New("make plan must be called before replication flow planner contains srcTenantID")
```
Let's make this an AssertionFailedf.

I could see changing the API of this type a little bit so that the usage was more clear.  Namely, rather than having the caller create a `&replicationFlowPlanner{}` and then calling makePlan on it. You could have a constructor that returns a flow planner with the srcTenantID and initial topology set.  Then you could have a method on the struct that was like `generateCandidatePlan()` that would assume it was constructed using the constructor. This function could also assume it was constructed using the constructor.

stevendanna commented on pkg/ccl/streamingccl/streamingest/testdata/span_configs:
```diff
@@ -0,0 +1,113 @@
+14400
+
+
+# FIGURE OUT HOW TO COMBINE THE SPANCONFIGS, CAUSING A DELETE
```
:D Let's make this a TODO.

Can we cause a delete by dropping the table gc.ttl on the table and then dropping the table?

msbutler replied:
Added a Todo. Chatting with Arul about this. It's harder than I initially thought.

stevendanna commented on pkg/ccl/streamingccl/streamingest/testdata/span_configs:
```diff
@@ -0,0 +1,113 @@
+ALTER TENANT ALL SET CLUSTER SETTING spanconfig.reconciliation_job.checkpoint_interval = '10ms'
+----
+
+# TODO(msbutler): trigger a full reconciliation job on the new app tenant after cutover, so that we
```
I'm a little surprised a full reconciliation doesn't happen on tenant startup.

msbutler replied:
Plan to address in a follow up PR

stevendanna commented on pkg/ccl/streamingccl/streamingest/ingest_span_configs.go:
```diff
@@ -0,0 +1,253 @@
+	"github.com/cockroachdb/errors"
+)
+
+type spanConfigIngestor struct {
```
Let's add a documentation comment to this struct describing what the spanConfigIngestor does and why.

stevendanna commented on pkg/ccl/streamingccl/streamingest/ingest_span_configs.go:
```diff
@@ -0,0 +1,253 @@
+	g.GoCtx(sub.Subscribe)
+	g.GoCtx(sc.checkForCutover)
+
+	return sc.consumeSpanConfigs(ctx, sub)
```
I may be missing something, but I would probably put this in the context group and then call g.Wait() here.  Otherwise, who bubbles up an error from sub.Subscribe?

If you go that route, you may need to arrange some way to shut down the cutover worker (unless we remove it). in the case of an error.  

How we do this in other places is to create a new context, grab a CancelFunc and then call the cancel func in close.

msbutler replied:
Good point! Refactored this a bit to make it work.

stevendanna reviewed (commented):
Thanks! Reviewed just the second commit.

msbutler reviewed (commented):
thanks for the quick review! I think I've addressed all your comments. I'm going to add one more unit test for the ingest_span_configs.go function tomorrow morning.

stevendanna commented on pkg/ccl/streamingccl/streamingest/ingest_span_configs.go:
```diff
@@ -0,0 +1,249 @@
+	group := ctxgroup.WithContext(ctx)
+	group.GoCtx(sub.Subscribe)
+	group.GoCtx(func(ctx context.Context) error {
+		defer func() {
+			if err := sc.client.Close(ctx); err != nil {
+				log.Errorf(ctx, "error on client.close(): %s", err)
+			}
+		}()
```
If you fancy, I think there is a closeAndLog(ctx, client) in this package

msbutler replied:
neat!

stevendanna reviewed (commented):
Thanks for addressing my feedback.

Overall, it looks like the next major TODO to handle is probably correctly handling full updates on resumption?

Let me know when this one is ready for final review.

msbutler committed:
streamingest: replicate span configs relevant to replicating tenant

This patch spawns a new goroutine on the stream ingestion job coordinator which
in turn creates a special span config client and subscription for replicating
span config updates. As updates come in, the coordinator will buffer updates
and deletes with the same source side commit timestamp and write them in a
transaction to the destination side system span configuration table when it
observes a new update with a newer timestamp.

The ingestion side assumes each replicated update is unique and in timestamp
order, which is enforced by the producer side logic built in #108356. This
assumption simplifies the destination side ingestion logic which must write
updates with the same source side transaction commit timestamp at the
same new timestamp on the destination side. This invariant ensures a span
configuration's target (i.e. the span that a configuration applies to) never
overlaps with any other span configuration target. Else, c2c would break the
span config reconciliation system.

Note that cutover will not revert any ingested span config updates, as we
wouldn't not want to issue revert range requests on an online system table in
the system tenant. That being said, a future PR will need to teach the
destination side application tenant to conduct a full span reconcilation job
immediately after cutover, which will safely revert any span config updates
that committed after the cutover timestamp.

A future PR will also improve the ingestion logic during an initial scan on
Resume. Because the job does not maintain a protected timestamp on the span
config table, the job may miss updates if it resumes after a long pause period.
So, the ingestor will need to buffer all initial scan updates and flush them in
one transaction along with a tenant span wide delete, in order to maintain the
span config table invariants.

Informs #106823

Release note: None

msbutler committed:
c2c: refactor stream ingestion distsql planner

This patch pushes the generation of the initial plan into a
`replicationFlowPlanner` new constructor. This refactor makes replanning logic
slightly easier to reason about and also ensures that the
`replicationFlowPlanner` gathers all necessary state (e.g. the srcTenantID)
before callers interact with.

Release Note: None

msbutler commented:
@stevendanna RFAL! 

> Overall, it looks like the next major TODO to handle is probably correctly handling full updates on resumption?
Correct

msbutler commented:
TFTR!

bors r=stevendanna

msbutler commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_dist.go:
```diff
@@ -192,28 +201,69 @@ func startDistIngestion(
 }
 
+// makeReplicationFlowPlanner creates a replicationFlowPlanner and the initial physical plan.
+func makeReplicationFlowPlanner(
```
yeah, i agree that this approach is inherently flawed. At some point we ought refactor this to make it easier to grab the `sourceTenantID` and streamAddresses.
